contrib/win32/win32compat/gss-sspi.c

```diff
@@ -1,1093 +1,1103 @@
+	GetSystemTime(&current_time_system);
+
+	/* acquire default cred handler if none specified */
+	CredHandle *pCredHandle = NULL;
+	if (claimant_cred_handle != NULL)
+		pCredHandle = &(claimant_cred_handle->credHandle);
+	
+	if (pCredHandle == NULL) {
+		static CredHandle cred_handle = { 0, 0 };
+		pCredHandle = &cred_handle;
+		if (cred_handle.dwLower == 0 && cred_handle.dwUpper == 0) {
+			TimeStamp expiry_cred;
+			if (SecFunctions->AcquireCredentialsHandleW(NULL, MICROSOFT_KERBEROS_NAME_W, SECPKG_CRED_OUTBOUND,
+			    NULL, NULL, NULL, NULL, &cred_handle, &expiry_cred) != SEC_E_OK)
+				goto done;
+		}
+	}
+
+	/* condition the string for windows */
+	if ((target_name_utf16 = utf8_to_utf16(target_name)) == NULL)
+		goto done;
+
+	if (wcsncmp(target_name_utf16, L"host@", wcslen(L"host@")) == 0) 
+		*wcschr(target_name_utf16, L'@') = L'/';
+
+	TimeStamp expiry;
+	LONG sspi_ret_flags = 0;
+	CtxtHandle out_context;
+
+	const SECURITY_STATUS status = SecFunctions->InitializeSecurityContextW(pCredHandle,
+		(*context_handle == GSS_C_NO_CONTEXT) ? NULL : *context_handle,
+		target_name_utf16, sspi_req_flags, 0, SECURITY_NATIVE_DREP, (no_input_buffer) ? NULL : &input_buffer,
+		0, (*context_handle != NULL) ? NULL : &out_context, &output_buffer, &sspi_ret_flags, (time_rec == NULL) ? NULL : &expiry);
+
+	/* check if error occurred */
+	if (status != SEC_E_OK && status != SEC_I_CONTINUE_NEEDED)
+		goto done;
+	
+	/* copy output token to output buffer */
+	output_token->length = output_buffer_token.cbBuffer;
+	if ((output_token->value = malloc(output_token->length)) == NULL)
+		goto done;
+
+	memcpy(output_token->value, output_buffer_token.pvBuffer, output_token->length);
+	SecFunctions->FreeContextBuffer(output_buffer_token.pvBuffer);
+
+	/* if requested, translate returned flags that are actually available */
+	if (ret_flags != NULL) {
+		*ret_flags = 0;
+		if (sspi_ret_flags & ISC_RET_MUTUAL_AUTH) 
+			*ret_flags |= GSS_C_MUTUAL_FLAG;
+		if (sspi_ret_flags & ISC_RET_CONFIDENTIALITY) 
+			*ret_flags |= GSS_C_CONF_FLAG;
+		if (sspi_ret_flags & ISC_RET_REPLAY_DETECT) 
+			*ret_flags |= GSS_C_REPLAY_FLAG;
+		if (sspi_ret_flags & ISC_RET_DELEGATE) 
+			*ret_flags |= GSS_C_DELEG_FLAG;
+		if (sspi_ret_flags & ISC_RET_INTEGRITY) 
+			*ret_flags |= GSS_C_INTEG_FLAG;
+		if (sspi_ret_flags & ISC_RET_SEQUENCE_DETECT) 
+			*ret_flags |= GSS_C_SEQUENCE_FLAG;
+	}
+
+	/* report if delegation was requested by not fulfilled */
+	if ((sspi_req_flags & ISC_REQ_DELEGATE) != 0 && (sspi_ret_flags & ISC_RET_DELEGATE) == 0)
+		debug("sspi delegation was requested but not fulfilled");
+
+	/* if requested, translate the expiration time to number of second */
+	if (time_rec != NULL) {
+		FILETIME current_time;
+		if (SystemTimeToFileTime(&current_time_system, &current_time) != 0)
+			*time_rec = (OM_uint32)(expiry.QuadPart - ((PLARGE_INTEGER)&current_time)->QuadPart) / 10000;
+		else
+			debug("gss_init_sec_context: SystemTimeToFileTime failed");

```

Comment:
error log instead of debug log.
capture the error information using GetLastError().